### PR TITLE
add permission boundary to addon role

### PR DESCRIPTION
**Pull request by aclevername**
### Description

https://github.com/weaveworks/eksctl/issues/3236

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [ ] Manually tested
- [ ] Made sure the title of the PR is a good description that can go into the release notes
- [ ] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by aclevername** [comment]
running the addons and CRUD integration tests as they should ensure I haven't broken anything

**Comment by aclevername** [comment]
One thing I noticed is that we don't support providing permission boundary in the `eksctl create iamserviceaccount` via a flag, so I kept it consistent here, not sure if that was intentional to begin with or not.

**Review by michaelbeaumont — COMMENTED** [review]
Does `AddAllResources` have the same behavior? I.e. does anything change for the addon stacks?

File: pkg/cfn/builder/iam.go [hunk 1]
```diff
@@ -182,164 +182,116 @@ func (n *NodeGroupResourceSet) addResourcesForIAM() error {
...
-
-	// TODO: declare output collector automatically when all stack builders migrated to our template package
-	rs.template.Outputs["Role1"] = cft.Output{
-		Value: cft.MakeFnGetAttString("Role1.Arn"),
-	}
-	rs.outputs = outputs.NewCollectorSet(map[string]outputs.Collector{
-		"Role1": func(v string) error {
-			rs.spec.Status = &api.ClusterIAMServiceAccountStatus{
-				RoleARN: &v,
-			}
```
[the remark this answers is not in the document]
Comment (aclevername) [reply pkg/cfn/builder/iam.go]: fixed

**Comment by aclevername** [comment]
> Does `AddAllResources` have the same behavior? I.e. does anything change for the addon stacks?

So I believe they have the same behaviour. Originally when addons was added I copied the IAMServiceAccount resource set and removed the permission boundary & name/namespace sub condition. Since then both of have come back, so AFAIK they are now identical, which makes sense as the addon is basically running `eksctl create iamserviceaccount` without the creation or annotation of the service account.

File: pkg/cfn/builder/iam.go [hunk 2]
```diff
@@ -182,172 +182,141 @@ func (n *NodeGroupResourceSet) addResourcesForIAM() error {
...
-func NewIAMRoleResourceSetWithAttachPolicyARNs(name, namespace, serviceAccount string, attachPolicyARNs []string, oidc *iamoidc.OpenIDConnectManager) *IAMRoleResourceSet {
-	return &IAMRoleResourceSet{
-		template:         cft.NewTemplate(),
-		attachPolicyARNs: attachPolicyARNs,
-		name:             name,
-		oidc:             oidc,
-		serviceAccount:   serviceAccount,
-		namespace:        namespace,
+// NewIAMRoleResourceSetForServiceAccount builds IAM Role stack from the give spec
+func NewIAMRoleResourceSetWithAttachPolicyARNs(name, namespace, serviceAccount, permissionsBoundary string, attachPolicyARNs []string, oidc *iamoidc.OpenIDConnectManager) *IAMRoleResourceSet {
```
Comment (michaelbeaumont) [inline pkg/cfn/builder/iam.go:228]: This seems to duplicate a lot from `WithAttachPolicy`, can they be cleaned up?

Comment (aclevername) [reply pkg/cfn/builder/iam.go:228]: :+1:

**Review by michaelbeaumont — APPROVED** [review]

**Pull request merged**